fixup rollover for powercap and msr drivers

cmcantalupo opened the pull request:
- Fixes #3914
- Also fixes rollover for MSR signals through read_signal interface (reuse RolloverGenerator)

cmcantalupo commented:
🤖 

Suggestions

    Consider initializing last_energy on first use to prevent a false overflow:
    C++

    if (overflow_count == 0 && last_energy == 0) {
        last_energy = current_energy;
    }

    Optionally, clarify in comments how thread-local storage is intended to be used.

Summary:
Your rollover logic is sound and should work correctly for unsigned counters. The main thing to double-check is initialization of last_energy to avoid a spurious overflow on the first sample, and to ensure that thread-local storage matches your usage patterns.

Let me know if you want a more detailed code suggestion or review of the surrounding context!

cmcantalupo committed:
Fix rollover issue with powercap sysfs interface

- Fixes #3914
- Add a generator for overflow
- Add sysfs attribute for rollover value
- Remove warning about rollover from geopmopt

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on libgeopmd/src/CpuinfoIOGroup.cpp:
```diff
@@ -184,11 +184,6 @@ namespace geopm
                             "not valid for CpuinfoIOGroup",
                             GEOPM_ERROR_INVALID, __FILE__, __LINE__);
         }
-        else if (domain_type != GEOPM_DOMAIN_BOARD) {
```
Seems the docu already had this set to CPU:
https://geopm.github.io/geopm_pio_cpuinfo.7.html

Since these were previously only allowed to be read at the BOARD level, we should test to make sure the other domains are working as expected re: aggregation.

cmcantalupo replied:
Yeah this check is *actually* the responsibility of PlatformIO.  It only ever makes requests of an IOGroup at the native domain for a signal or control.  These exceptions were both incorrect (board vs CPU) and also redundant.

cmcantalupo replied:
Checked, and LGTM, all the signals are frequencies

bgeltz commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -105,7 +107,18 @@ namespace geopm
+        , m_rollover_factor(0.0)
     {
+        try {
+            std::string factor_path = POWERCAP_DIRECTORY;
+            factor_path += "/intel-rapl/intel-rapl:0/max_energy_range_uj";
```
Can't you get at this with ``M_PROPERTIES``, ``name_to_property_key()``, and the newly defined ``MAX_ENERGY_RANGE`` instead of duplicating the JSON data here?

bgeltz commented on libgeopmd/src/RolloverGenerator.cpp:
```diff
@@ -0,0 +1,36 @@
+{
+    RolloverGenerator::RolloverGenerator(int bit_width)
+        : m_last_value(0.0)
+        , m_rollover_factor(pow(2, bit_width))
```
Why set ``m_rollover_factor`` in the initializer list, and also provide the ``set_factor()`` method?  Why both?

cmcantalupo replied:
The constructor takes a bit width, but in the case of powercap, we only get the max value in microjoules.  We could definitly consider having the constructor take the rollover factor directly.

cmcantalupo commented on libgeopmd/src/RolloverGenerator.cpp:
```diff
@@ -0,0 +1,36 @@
+{
+    RolloverGenerator::RolloverGenerator(int bit_width)
+        : m_last_value(0.0)
+        , m_rollover_factor(pow(2, bit_width))
```
doesn't work for bit_width == 64, but yet another reason to put get rid of the bit_width constructor.

cmcantalupo commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -150,10 +162,18 @@ namespace geopm
+            rollover_ptr = std::make_shared<RolloverGenerator>(64);
+            rollover_ptr->set_factor(m_rollover_factor);
+        }
+        return [scaling_factor, rollover_ptr](const std::string &content) {
```
we want a copy...rollover_ptr will go out of scope. I'm not sure if we can std::move() in the capture?

cmcantalupo committed:
Prefer powercap over default for optimizer

- Defaults map to msr interface in 3.2.0
- Relates to #3934
- Relates to #3914

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move pointer in capture to avoid copy

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix rollover when using msr read_signal() interface

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up error reporting

- Adjust logic for 2^64 calculation
- Remove unneeded includes

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on libgeopmd/src/RolloverGenerator.cpp:
```diff
@@ -0,0 +1,33 @@
+        if (m_last_value > value) {
+            ++m_num_overflow;
+        }
+        double rollover = m_rollover_factor * m_num_overflow;
```
This could be inside the if statement above as a different member variable to avoid running this calculation during every update() call.

bgeltz commented on libgeopmd/src/PowercapSysfsDriver.hpp:
```diff
@@ -6,11 +6,15 @@
 #ifndef POWERCAPSYSFSDRIVER_HPP_INCLUDE
 #define POWERCAPSYSFSDRIVER_HPP_INCLUDE
 
+#include <memory>
+#include <vector>
```
```suggestion
#include <memory>
#include <map>
```

bgeltz commented on libgeopmd/test/MSRFieldSignalTest.cpp:
```diff
@@ -143,19 +143,19 @@ TEST_F(MSRFieldSignalTest, read_overflow)
     result = sig->read();
```
Comment on 139 should be removed.

bgeltz commented on libgeopmd/test/RolloverGeneratorTest.cpp:
```diff
@@ -0,0 +1,98 @@
+
+#include "geopm/SaveControl.hpp"
+
+#include <iostream>
```
```suggestion
```

bgeltz commented on libgeopmd/test/RolloverGeneratorTest.cpp:
```diff
@@ -0,0 +1,98 @@
+{
+    auto gen = RolloverGenerator();
+    gen.set_factor(pow(2, 2));
+    for (int idx = 0; idx < 4; ++idx) {
```
```suggestion
    for (int idx = 0; idx < 8; ++idx) {
```

bgeltz commented on libgeopmd/test/RolloverGeneratorTest.cpp:
```diff
@@ -0,0 +1,98 @@
+    double rollover_factor = static_cast<double>(UINT32_MAX) + 1.0;
+    EXPECT_EQ(rollover_factor, gen.update(0));
+    EXPECT_EQ(rollover_factor + 1.0, gen.update(1));
+    EXPECT_NEAR(2.0 * rollover_factor + 1.0, gen.update(0), 1);
```
Could this be EXPECT_EQ?

cmcantalupo committed:
Make RolloverGenerator more efficient

- Fix unit test
- Remove unneeded includes

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Improve comments in test

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -150,10 +174,18 @@ namespace geopm
+            rollover_ptr = std::make_shared<RolloverGenerator>();
+            rollover_ptr->set_factor(m_rollover_factor);
+        }
+        return [scaling_factor, rollover_ptr = std::move(rollover_ptr)](const std::string &content) {
```
there is one lambda per signal and per domain

cmcantalupo commented on libgeopmd/test/RolloverGeneratorTest.cpp:
```diff
@@ -0,0 +1,96 @@
+    EXPECT_EQ(rollover_factor + 1.0, gen.update(1));
+    double expected = 2.0 * rollover_factor;
+    // IEEE 754 double precision floating point mantissa bit width is 51
+    double error = 1 << (64 - 51);
```
This is a test....

cmcantalupo commented on libgeopmd/src/MSRFieldSignal.cpp:
```diff
@@ -75,10 +79,7 @@ namespace geopm
-                    ++num_overflow;
-                }
-                result = subfield + ((m_subfield_max + 1.0) * num_overflow);
+                result = m_rollover_gen->update(subfield);
```
Copilot seems to be confused.  There is one MSRFieldSignal per subfield per msr per cpu

bgeltz commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -105,7 +108,28 @@ namespace geopm
+                                    GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+         }
+         try {
+             m_rollover_factor = 1e-6 * std::stoll(contents);
```
A comment explaining where the magic ``1e-6`` came from would be helpful.  Is it possible to put a link to the upstream docu for this sysfs entry here?
Also, should this be a constexpr instead of hard coded here?

cmcantalupo committed:
Avoid magic number, use data

- Fix some whitespace

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on libgeopmd/src/PowercapSysfsDriver.cpp:
```diff
@@ -105,7 +108,35 @@ namespace geopm
     {
+        std::string signal_name = "POWERCAP::CPU_MAX_ENERGY_RANGE";
+        std::string factor_path = attribute_path(signal_name, 0);
+        std::string property_key = name_to_property_key(signal_name);
```
```suggestion
        std::string property_key = name_to_property_key(signal_name);
        std::string contents;
```
